<!doctype html><html lang=en-us dir=ltr itemscope itemtype=http://schema.org/Article><head><meta charset=utf-8><meta name=viewport content="height=device-height,width=device-width,initial-scale=1,minimum-scale=1"><meta name=generator content="Hugo 0.128.0"><meta name=generator content="Relearn 6.0.0"><meta name=robots content="noindex, nofollow, noarchive, noimageindex"><meta name=description content="Exception Handling As was mentioned in the previous section, various problems can occur when doing file I/O. Some of these problems include:
Trying to write to a read-only file. Trying to access a file that is locked by another process. Accessing an external drive that becomes disconnected. Note that some of these issues are beyond the programmer’s control, while others may be tedious for the programmer to check. When one of these problems prevents an I/O operation from completing successfully, an exception is thrown."><meta name=author content="Rod Howell and Josh Weese"><meta name=twitter:card content="summary"><meta name=twitter:title content="Exception Handling :: Data Structures in C#"><meta name=twitter:description content="Exception Handling As was mentioned in the previous section, various problems can occur when doing file I/O. Some of these problems include:
Trying to write to a read-only file. Trying to access a file that is locked by another process. Accessing an external drive that becomes disconnected. Note that some of these issues are beyond the programmer’s control, while others may be tedious for the programmer to check. When one of these problems prevents an I/O operation from completing successfully, an exception is thrown."><meta property="og:url" content="https://textbooks.cs.ksu.edu/cis300/io/exceptions/embed.html"><meta property="og:site_name" content="Data Structures in C#"><meta property="og:title" content="Exception Handling :: Data Structures in C#"><meta property="og:description" content="Exception Handling As was mentioned in the previous section, various problems can occur when doing file I/O. Some of these problems include:
Trying to write to a read-only file. Trying to access a file that is locked by another process. Accessing an external drive that becomes disconnected. Note that some of these issues are beyond the programmer’s control, while others may be tedious for the programmer to check. When one of these problems prevents an I/O operation from completing successfully, an exception is thrown."><meta property="og:locale" content="en-us"><meta property="og:type" content="article"><meta property="article:section" content="Input/Output"><meta property="article:modified_time" content="2023-05-31T10:25:43-05:00"><meta itemprop=name content="Exception Handling :: Data Structures in C#"><meta itemprop=description content="Exception Handling As was mentioned in the previous section, various problems can occur when doing file I/O. Some of these problems include:
Trying to write to a read-only file. Trying to access a file that is locked by another process. Accessing an external drive that becomes disconnected. Note that some of these issues are beyond the programmer’s control, while others may be tedious for the programmer to check. When one of these problems prevents an I/O operation from completing successfully, an exception is thrown."><meta itemprop=dateModified content="2023-05-31T10:25:43-05:00"><meta itemprop=wordCount content="1074"><title>Exception Handling :: Data Structures in C#</title>
<link href=/cis300/css/fontawesome-all.min.css?1723839632 rel=stylesheet media=print onload='this.media="all",this.onload=null'><noscript><link href=/cis300/css/fontawesome-all.min.css?1723839632 rel=stylesheet></noscript><link href=/cis300/css/nucleus.css?1723839632 rel=stylesheet><link href=/cis300/css/auto-complete.css?1723839632 rel=stylesheet media=print onload='this.media="all",this.onload=null'><noscript><link href=/cis300/css/auto-complete.css?1723839632 rel=stylesheet></noscript><link href=/cis300/css/perfect-scrollbar.min.css?1723839632 rel=stylesheet><link href=/cis300/css/fonts.css?1723839632 rel=stylesheet media=print onload='this.media="all",this.onload=null'><noscript><link href=/cis300/css/fonts.css?1723839632 rel=stylesheet></noscript><link href=/cis300/css/theme.css?1723839632 rel=stylesheet><link href=/cis300/css/theme-light-theme.css?1723839632 rel=stylesheet id=R-variant-style><link href=/cis300/css/chroma-relearn-light.css?1723839632 rel=stylesheet id=R-variant-chroma-style><link href=/cis300/css/variant.css?1723839632 rel=stylesheet><link href=/cis300/css/print.css?1723839632 rel=stylesheet media=print><script>window.relearn=window.relearn||{},window.relearn.relBasePath="../..",window.relearn.relBaseUri="../../..",window.relearn.absBaseUri="https://textbooks.cs.ksu.edu/cis300",window.index_js_url="/cis300/index.search.js",window.variants&&variants.init(["light-theme"]),window.T_Copy_to_clipboard=`Copy to clipboard`,window.T_Copied_to_clipboard=`Copied to clipboard!`,window.T_Copy_link_to_clipboard=`Copy link to clipboard`,window.T_Link_copied_to_clipboard=`Copied link to clipboard!`,window.T_Reset_view=`Reset view`,window.T_View_reset=`View reset!`,window.T_No_results_found=`No results found for "{0}"`,window.T_N_results_found=`{1} results found for "{0}"`</script><link href=/cis300/css/custom.css?1723839632 rel=stylesheet></head><body class="mobile-support embed disableInlineCopyToClipboard" data-url=/cis300/io/exceptions/embed.html><div id=R-body class=default-animation><div id=R-body-overlay></div><div id=R-main-overlay></div><main id=R-body-inner class="highlightable default" tabindex=-1><div class=flex-block-wrapper><article class=default><h2 id=exception-handling>Exception Handling</h2><p>As was mentioned in <a href=/cis300/io/simple-text-file/>the previous
section</a>, various
problems can occur when doing file I/O. Some of these problems include:</p><ul><li>Trying to write to a read-only file.</li><li>Trying to access a file that is locked by another process.</li><li>Accessing an external drive that becomes disconnected.</li></ul><p>Note that some of these issues are beyond the programmer&rsquo;s control,
while others may be tedious for the programmer to check. When one of these problems
prevents an I/O operation from completing successfully, an exception is
thrown. This section discusses how to handle such exceptions gracefully,
without terminating the program.</p><div class="box notices cstyle tip"><div class=box-label><i class="fa-fw fas fa-lightbulb"></i> Tip</div><div class=box-content><p><a href=/cis300/io/dialogs/file-dialogs/>File dialogs</a>
can be quite helpful in avoiding some of these exceptions, as they
can reject improper selections by the user.</p></div></div><p>The mechanism used to handle exceptions is the <strong>try</strong>-<strong>catch</strong>
construct. In its simplest form, it looks like:</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-C# data-lang=C#><span class=line><span class=cl><span class=k>try</span>
</span></span><span class=line><span class=cl><span class=p>{</span>
</span></span><span class=line><span class=cl>    
</span></span><span class=line><span class=cl>    <span class=c1>// Block of code that might throw an exception</span>
</span></span><span class=line><span class=cl>    
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl><span class=k>catch</span>
</span></span><span class=line><span class=cl><span class=p>{</span>
</span></span><span class=line><span class=cl>    
</span></span><span class=line><span class=cl>    <span class=c1>// Code to handle the exception</span>
</span></span><span class=line><span class=cl>    
</span></span><span class=line><span class=cl><span class=p>}</span></span></span></code></pre></div><p>If we are concerned about exceptions thrown while doing I/O, we would
include the I/O and anything dependent on it within the <strong>try</strong>-block.
If at any point within this block an exception is thrown, control
immediately jumps to the <strong>catch</strong>-block. Here, we would place code to
handle the exception - for example, displaying a message to the user.</p><p>Suppose, for example, that we want to count the number of upper-case
letters in a file whose name is in the <strong>string</strong> referenced by
<code>fileName</code>. We could use the following code:</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-C# data-lang=C#><span class=line><span class=cl><span class=k>try</span>
</span></span><span class=line><span class=cl><span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=kt>string</span> <span class=n>contents</span> <span class=p>=</span> <span class=n>File</span><span class=p>.</span><span class=n>ReadAllText</span><span class=p>(</span><span class=n>fileName</span><span class=p>);</span>
</span></span><span class=line><span class=cl>    <span class=kt>int</span> <span class=n>count</span> <span class=p>=</span> <span class=m>0</span><span class=p>;</span>
</span></span><span class=line><span class=cl>    <span class=k>foreach</span> <span class=p>(</span><span class=kt>char</span> <span class=n>c</span> <span class=k>in</span> <span class=n>contents</span><span class=p>)</span>
</span></span><span class=line><span class=cl>    <span class=p>{</span>
</span></span><span class=line><span class=cl>        <span class=k>if</span> <span class=p>(</span><span class=kt>char</span><span class=p>.</span><span class=n>IsUpper</span><span class=p>(</span><span class=n>c</span><span class=p>))</span>
</span></span><span class=line><span class=cl>        <span class=p>{</span>
</span></span><span class=line><span class=cl>            <span class=n>count</span><span class=p>++;</span>
</span></span><span class=line><span class=cl>        <span class=p>}</span>
</span></span><span class=line><span class=cl>    <span class=p>}</span>
</span></span><span class=line><span class=cl>    <span class=n>MessageBox</span><span class=p>.</span><span class=n>Show</span><span class=p>(</span><span class=s>&#34;The file contains &#34;</span> <span class=p>+</span> <span class=n>count</span> <span class=p>+</span> <span class=s>&#34; upper-case letters.&#34;</span><span class=p>);</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl><span class=k>catch</span>
</span></span><span class=line><span class=cl><span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>MessageBox</span><span class=p>.</span><span class=n>Show</span><span class=p>(</span><span class=s>&#34;An exception occurred.&#34;</span><span class=p>);</span>
</span></span><span class=line><span class=cl><span class=p>}</span></span></span></code></pre></div><div class="box notices cstyle note"><div class=box-label><i class="fa-fw fas fa-exclamation-circle"></i> Note</div><div class=box-content><p>See the section, &ldquo;<a href=/cis300/appendix/syntax/foreach/>The <strong>foreach</strong>
Statement</a>&rdquo; for an
explanation of <strong>foreach</strong> loops. The
<a href="https://learn.microsoft.com/en-us/dotnet/api/system.char.isupper?view=net-6.0#system-char-isupper(system-char)" rel=external target=_blank><strong>char.IsUpper</strong></a>
method returns a <strong>bool</strong> indicating whether the given <strong>char</strong> is an
upper-case letter in some alphabet.</p></div></div><p>We should always include within the <strong>try</strong>-block all of the code that
depends on what we want to read. Consider what would happen, for
example, if we tried to move the statement,</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-C# data-lang=C#><span class=line><span class=cl><span class=n>MessageBox</span><span class=p>.</span><span class=n>Show</span><span class=p>(</span><span class=s>&#34;The file contains &#34;</span> <span class=p>+</span> <span class=n>count</span> <span class=p>+</span> <span class=s>&#34; upper-case letters.&#34;</span><span class=p>);</span></span></span></code></pre></div><p>outside the <strong>try</strong>-<strong>catch</strong>. First, we would have a syntax error
because the variable <code>count</code> is declared within the <strong>try</strong>-block, and
hence cannot be used outside of it. We could fix this error by declaring
and initializing <code>count</code> prior to the <strong>try</strong> statement. The resulting
code would compile and run, but consider what happens if an exception is
thrown during the reading of the file. Control immediately jumps to the
<strong>catch</strong>-block, where the message, &ldquo;An exception occurred.&rdquo;, is
displayed. After that, assuming we have made these changes to the above
code, control continues on past the <strong>catch</strong>-block to the code to
display the results. Because the file was not successfully read, it
really doesn&rsquo;t make any sense to do this. The code given above, however,
displays a result only if the result is successfully computed;
otherwise, the exception message is displayed.</p><p>In the above example, the message, &ldquo;An exception occurred.&rdquo;, isn&rsquo;t very
helpful to the user. It gives no indication of what the problem is. In
order to be able to provide more information to the user, we need more
information regarding the nature of the exception. The way we do this is
to use some additional code on the <strong>catch</strong> statement:</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-C# data-lang=C#><span class=line><span class=cl><span class=k>catch</span> <span class=p>(</span><span class=n>Exception</span> <span class=n>ex</span><span class=p>)</span></span></span></code></pre></div><p>The word <strong>Exception</strong> above is a type. Every exception in C# is a
subtype of the <strong>Exception</strong> class. In this form of the <strong>catch</strong>
statement, we can include any subtype of <strong>Exception</strong>, including
<strong>Exception</strong> itself, as the first word within the parentheses. The
second word is a new variable name. One effect of this parenthesized
part is to declare this variable to be of the given type; i.e., <code>ex</code> is
of type <strong>Exception</strong>, and may be used within the <strong>catch</strong> block.</p><p>This form of the <strong>catch</strong> statement will catch any exception that can
be treated as the given type. If we use the type, <strong>Exception</strong>, as
above, the <strong>catch</strong>-block will still catch any exception. In addition,
the variable defined within the parentheses will refer to that
exception. Thus, the parenthesized part of this statement behaves much
like a parameter list, giving us access to the exception that was
thrown. Having the exception available to examine, we can now give more
meaningful feedback to the user. One rather crude way of doing this is
to use the exception&rsquo;s
<a href="https://learn.microsoft.com/en-us/dotnet/api/system.object.tostring?view=net-6.0#system-object-tostring" rel=external target=_blank><strong>ToString</strong></a>
method to convert it to a <strong>string</strong>
representation, which can then be displayed to the user; for example,</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-C# data-lang=C#><span class=line><span class=cl><span class=k>catch</span> <span class=p>(</span><span class=n>Exception</span> <span class=n>ex</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>MessageBox</span><span class=p>.</span><span class=n>Show</span><span class=p>(</span><span class=n>ex</span><span class=p>.</span><span class=n>ToString</span><span class=p>());</span>
</span></span><span class=line><span class=cl><span class=p>}</span></span></span></code></pre></div><p>Replacing the <strong>catch</strong>-block in the earler example with this
<strong>catch</strong>-block might result in the following message:</p><img src=exception-in-messagebox.png alt="A MessageBox displaying an exception." style=zoom:67%><p>While this message is not something we would want to show to an end
user, it does provide helpful debugging information, such as the
exception thrown and the line that threw the exception.</p><div class="box notices cstyle tip"><div class=box-label><i class="fa-fw fas fa-lightbulb"></i> Tip</div><div class=box-content><p>Every object in C# has a
<a href="https://learn.microsoft.com/en-us/dotnet/api/system.object.tostring?view=net-6.0#system-object-tostring" rel=external target=_blank><strong>ToString</strong></a>
method. Hence, we can convert an instance of any type to <strong>string</strong> by
using its <strong>ToString</strong> method. This method will always return a
<strong>string</strong>, but depending on the original type, this <strong>string</strong> may or
may not be useful. For example, because there is no particularly meaningful way to
convert a <strong>Form</strong> to a <strong>string</strong>, its <strong>ToString</strong> method is not
especially useful.</p></div></div><p>A single <strong>try</strong>-block can have more than one <strong>catch</strong>-block. In such a
case, whenever an exception occurs within the <strong>try</strong>-block, control is
transferred to the first <strong>catch</strong>-block that can catch that particular
exception. For example, we can set up the following construct:</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-C# data-lang=C#><span class=line><span class=cl><span class=k>try</span>
</span></span><span class=line><span class=cl><span class=p>{</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>    <span class=c1>// Code that may throw an exception</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl><span class=k>catch</span> <span class=p>(</span><span class=n>DirectoryNotFoundException</span> <span class=n>ex</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>{</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>    <span class=c1>// Code to handle a DirectoryNotFoundException</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl><span class=k>catch</span> <span class=p>(</span><span class=n>FileNotFoundException</span> <span class=n>ex</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>{</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>    <span class=c1>// Code to handle a FileNotFoundException</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl><span class=k>catch</span> <span class=p>(</span><span class=n>Exception</span> <span class=n>ex</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>{</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>    <span class=c1>// Code to handle any other exception</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=p>}</span></span></span></code></pre></div><p>If we don&rsquo;t need access to the exception itself in order to handle it,
but only need to know what kind of exception it is, we can leave off the
variable name in the <strong>catch</strong> statement. For example, if we are trying
to read from a file whose name is referenced by the <strong>string</strong>
<code>fileName</code>, we might handle a <strong>FileNotFoundException</strong> as follows:</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-C# data-lang=C#><span class=line><span class=cl><span class=k>catch</span> <span class=p>(</span><span class=n>FileNotFoundException</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>{</span>
</span></span><span class=line><span class=cl>    <span class=n>MessageBox</span><span class=p>.</span><span class=n>Show</span><span class=p>(</span><span class=s>&#34;Could not find the file &#34;</span> <span class=p>+</span> <span class=n>fileName</span><span class=p>);</span>
</span></span><span class=line><span class=cl><span class=p>}</span></span></span></code></pre></div><div class="box notices cstyle warning"><div class=box-label><i class="fa-fw fas fa-exclamation-triangle"></i> Warning</div><div class=box-content><p>Don&rsquo;t use exception handling (i.e., <code>try</code>-<code>catch</code>) to handle cases that are expected to occur under normal conditions. In such cases, use an <code>if</code>-statement instead. Not only is this better style, but it is also more efficient.</p></div></div><footer class=footline></footer></article></div></main></div><script src=/cis300/js/clipboard.min.js?1723839632 defer></script><script src=/cis300/js/perfect-scrollbar.min.js?1723839632 defer></script><script src=/cis300/js/theme.js?1723839632 defer></script><script src=/cis300/js/embed-iframe.js?1723839632 defer></script></body></html>